### PR TITLE
perf: optimize poolKey.toId() (skip abi.encode)

### PR DESCRIPTION
**Pull request by adhusson**
## Related Issue
Which issue does this pull request resolve? None

## Description of changes
Save ~250 gas for every call to `poolKey.toId()` by skipping `abi.encode`. Added a test to check that the result is the same.

### CODE REVIEW

File: contracts/types/PoolId.sol
Comment (MerlinEgalite) [inline contracts/types/PoolId.sol:12]:
```suggestion
            poolId := keccak256(poolKey, mul(32, 5))
```

File: test/foundry-tests/types/PoolKey.t.sol
Comment (MerlinEgalite) [inline test/foundry-tests/types/PoolKey.t.sol:11]:
```suggestion
    using PoolIdLibrary for PoolKey;
    
    function testPoolIdCreation(PoolKey memory poolKey) public {
        assertEq(keccak256(abi.encode(poolKey)), PoolId.unwrap(poolKey.toId()));
    }
```

**Review by MerlinEgalite — APPROVED** [review]
Nice finding ser!

File: contracts/types/PoolId.sol
Comment (ewilz) [inline contracts/types/PoolId.sol:12]: maybe a code comment above the assembly block could be nice to save a maintainer/reader a minute
```
// PoolId.wrap(keccak256(abi.encode(poolKey)));
```

File: contracts/types/PoolId.sol
Comment (ewilz) [inline contracts/types/PoolId.sol:11]: also maybe just stick with `/// @solidity memory-safe-assembly` convention for now to be consistent with the rest of the repo. we might change these later...

https://github.com/Uniswap/v4-core/blob/e112589c1eb264e74e8626f6b91fbf4d8ba54639/contracts/types/BalanceDelta.sol#L10

**Review by ewilz — CHANGES_REQUESTED** [review]
some smol comments, + I think I just merged something that created some conflicts for you, shoudl be minimal!

File: contracts/types/PoolId.sol
Comment (adhusson) [reply contracts/types/PoolId.sol:14]: This and next comment addressed in 0f6c44850ae5eb68078cb5308940b71d6ff11b03!

File: contracts/types/PoolId.sol
Comment (MerlinEgalite) [inline contracts/types/PoolId.sol:10]:
```suggestion
    /// @notice Returns value equal to keccak256(abi.encode(poolKey))
```

To match pattern of the overall the codebase

**Review by MerlinEgalite — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Review by Georgezhang714 — APPROVED** [review]

**Comment by ewilz** [comment]
looks like the linter just needs to be run!

**Comment by adhusson** [comment]
> looks like the linter just needs to be run!
Sorry I missed that, looks like this needs an approval after I updated the branch!

**Review by MerlinEgalite — APPROVED** [review]

**Comment by hensha256** [comment]
Hi! So sorry for the delay in merging this. Please can you merge in main? Then I can approve and we can merge :)

**Review by snreynolds — DISMISSED** [review]
Hey! In general think this is a great change. Could you update to main - would love to get this in as soon as possible! Thanks

**Comment by snreynolds** [comment]
> Hey! In general think this is a great change. Could you update to main - would love to get this in as soon as possible! Thanks

Also once you update to main, just re-request me so I can be notified to give you the final approval :)

**Comment by adhusson** [comment]
@snreynolds Sorry I did not understand the "I will need you to open this into a temp branch" part -- I tried opening a [new PR](https://github.com/Uniswap/v4-core/pull/544) but it did not run linters either.

**Comment by snreynolds** [comment]
> @snreynolds Sorry I did not understand the "I will need you to open this into a temp branch" part -- I tried opening a [new PR](https://github.com/Uniswap/v4-core/pull/544) but it did not run linters either.

Can you open the PR into this branch ? https://github.com/Uniswap/v4-core/tree/temp/optimize-toId

**Comment by snreynolds** [comment]
We have to merge your commits into a temporary branch so that we can reopen that temp branch on main so the CI will run. Afaik we don't have a way for CI to run on an outside contributors branch and this is the only way I know how to get around that

**Comment by adhusson** [comment]
Got it, I think #550 is what you want.

**Pull request closed without merging**
